src/common/utils.py

```diff
@@ -1,21 +1,33 @@
-
-from typing import List, Dict, Tuple, Set, Union, Any, Type, Callable
+from __future__ import annotations
```

Comment:
This file is very large and contains lots of disjoint utilities. It may be a good idea to separate them out. Tons of work for very little payoff, I know, but it would improve the readability of the codebase and make it easier to develope on in the future.